### ISSUE TITLE
Feature Request : PathNorm and PathSGD

### ISSUE THREAD

**Issue by AshishBora** [tensorflow/tensorflow#8911]
PathSGD was introduced in [this](https://arxiv.org/abs/1506.02617) paper. Is there existing support for this? If not, this is a feature request for:
- PathNorm computation (Equation 5 in the paper)
- PathSGD using the PathNorm

For the first part, the interface can be to provide a function `path_norm(a, b, p=2)` where `a` and `b` are tensors, `p` is a scalar. The function returns the p-PathNorm for the "path" between the tensors `a` and `b` (assuming that `b` depends on `a` and some weights. If not, there can be an exception or simply return `0`).

**Comment by aselle** [comment]
I suspect there is not (given that I couldn't find a google search result for it), so I am marking it as contributions welcome. It is not something we are actively working on right now. Thanks1

**Comment by alisidd** [comment]
@AshishBora Do you wish to work on this, or can I?

**Comment by siryog90** [comment]
@alisidd @AshishBora Still interested in implementing this?

**Comment by AshishBora** [comment]
@alisidd @siryog90 I don't expect to have time to work on this anytime soon. So if one you wants to take this up, by all means go ahead!

**Comment by alisidd** [comment]
@AshishBora @siryog90 Yeah, I can work on it.

**Comment by siryog90** [comment]
@alisidd you want an extra pair of hands on this?

**Comment by alisidd** [comment]
@siryog90 Sure, that'd be great! I sent you an invite to my repo (in the pathsgd branch). I likely won't start until Wednesday, but you can start without me if you wish

**Comment by siryog90** [comment]
@alisidd I am thinking we implement pathNorm as a standalone C++ op first. I like OP's suggested interface. Thoughts? 
P.S. I am currently digging through the tensorflow graph framework (C++) to see if there are any useful abstractions for identifying a directed path between Tensors.

**Comment by alisidd** [comment]
@siryog90 Did you delete your message on implementing it as a C++ kernel? I can't see it anymore

**Comment by siryog90** [comment]
@alisidd I did delete  my previous message, because I think the following makes more sense. 
I am starting the Eigen (C++) implementation of the L_p-path regularizer tonight, and hopefully a basic python port for layers/regularizers.py
Does that sound ok?

**Comment by alisidd** [comment]
@siryog90 I don't know if a C++ kernel will get accepted due to #8405 

Any thoughts, @aselle?

**Comment by siryog90** [comment]
@alisidd Oh wow..
@aselle I am also working on an implementation of a Line Search optimizer. Should I not write the kernel in C++?

**Comment by aselle** [comment]
@lukaszkaiser in #8405, recommends using python if possible. I would agree in general. @lukaszkaiser, how would you rate the importance of these optimizer and if they should be written in C++.

**Comment by lukaszkaiser** [comment]
I'm not sure exactly, we should probably ask Martin Wicke. What I mean is: we definitely don't want to loose 3x performance of an important optimizer just because of "python is easier to maintain". But it is easier to maintain, so we probably don't want to add C++ kernels that we'll have to maintain for just 10% performance gain. Where the balance lies exactly is harder to say, it's a tradeoff. If the kernels are small and readable and the performance gains are big, I guess we'll want them, and vice versa. What are the performance gains from having separate kernels in this case?

**Comment by siryog90** [comment]
@aselle @lukaszkaiser For the L_p-path regularizer as a standalone module, the Eigen kernel can be small and quite readable, similar to this https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/l2loss_op.h
I think we can write the Path-SGD in Python for now, let people play with it, and if it shows promise, consider a C++ kernel?
As far as the LineSearch, it is unlike any other optimizer, in fact, it may need a new base class altogether, with a virtual StepDirection function, so maybe, for something this ambitious, C++ kernels would be the way to go.

**Comment by siryog90** [comment]
@aselle @lukaszkaiser @alisidd  Related to the topic, finally got a chance to implement the regularizer from the paper. It passed the basic tests. Before I make a pull request, I wanted to ask if I should implement it's gradient in C++ as, it seems, all the gradients of regularizers are implemented as C++ ops.
Thanks in advance.

**Comment by alisidd** [comment]
@siryog90 could you push it into github so I could work on it too?

**Comment by siryog90** [comment]
@alisidd Ali I just pushed it to my fork and sent you an invite.
One thing you can do right off the start is take a look at the modified `apply_regularizer` function. Non linear regularizers will require all the weights at once, and so, for testing, what I have done is pass the whole list of the weights to l2_path regularizer, since a list of weights is what `apply_regularizer` gets as an input. However, in order to implement the gradient of the regularizer, it would be much more natural had it accepted a single Tensor as input and not a list. So maybe you modify the `apply_gradients` so that it packs the list of tensors into a one higher rank tensor and passes that to the regularizer. Everything in side of l2_path itself should still work fine. Once this is done, maybe write the gradient for the l2_path? https://github.com/tensorflow/tensorflow/blob/904edee4456a61d50d5b1ffe9858a7772acc423e/tensorflow/python/ops/nn_grad.py

Don't hesitate to shoot a message if you have any questions. My email is sergeyyerevan@gmail.com

**Comment by AshishBora** [comment]
The path norm computes a sum of products over edge paths. Thus we should be multiplying individual weights, not the L2 norm of weight matrices. So I think [this](https://github.com/siryog90/tensorflow/commit/1d46853fc81ebf3c8f816e26055b574924ac26ff) implementation is not computing the path norm correctly.

For example, the path norm will never have interaction terms between two edges in consecutive layers that are not incident at the same neuron. But multiplying L2 norm of weights introduces those terms.

One way to compute the `Lp`-PathNorm can be this: 

- construct new weight matrices `W' = tf.pow(tf.abs(W), p)`.
- Use these in the same network architecture as the original.
- Give all ones tensor as input and pass it through the network to get some output.
- Compute the sum of all outputs and take that raise to `1/p`. This is the `Lp`-PathNorm.

**Comment by siryog90** [comment]
@alisidd @AshishBora Yup, definitely should not be introducing those terms. I will get on this tonight, but if you want to give it a go now @alisidd go for it

**Comment by alisidd** [comment]
@siryog90 I understood what he meant, I'm pushing it in a few mins

**Comment by AshishBora** [comment]
@siryog90 

Yes, we should replace the original weight matrices with ones raised to power `p`.

It doesn't matter if you use the original network or the one without non-linearities. 

Since the input (all ones tensor) is positive, and absolute value of (weights)^p is positive, each neuron will have positive activation and the ReLU wouldn't change anything.
This is because of the following:


First, PathNorm makes sense for only positive homogeneous non-linearities. i.e. those that satisfy `g(ax) = ag(x)` for all `x` and for all `a >= 0`. ReLU (and its friends like LeakyReLU etc.) satisfy this.



In addition if we also have g(x) = x for x>0, 


The non-linearities are all assumed to be ReLU

. (

**Comment by AshishBora** [comment]
@siryog90 

Yes, I think we should be using the original network with weights replacing weights with ones raised to power `p`.

It does not matter whether you use the original network or create a new one with identities. This is because the input (all ones tensor) is positive, and absolute value of (weights)^p is positive, each neuron will have positive activation and the ReLU wouldn't change anything.

More generally the PathNorm makes sense for any positive homogeneous non-linearity i.e. if 

`g(ax) = ag(x) for all x, and all a >= 0`.

This technique of passing all ones tensor will work with a minor modification accounting for the slope on the positive part for any such non-linearity. But for a first pass, we can focus on ReLUs and worry about this later.

**Comment by alisidd** [comment]
@AshishBora Could you look at the following and let me know if it is what you had in mind?

https://github.com/alisidd/tensorflow/commit/c8c85fc545a26fb944b47d5b00af7f5f212ab730

**Comment by alisidd** [comment]
@AshishBora Could you look at the following and let me know if it is what you had in mind?

https://github.com/alisidd/tensorflow/blob/patch-1/tensorflow/contrib/layers/python/layers/regularizers.py

It takes the path norm of the path vector

**Comment by siryog90** [comment]
@alisidd I may be missing something in your code, but I think he meant just sending an input tensor of ones down the network. What comes out will be the tensor we want. 
If we set `p` to 2, then both the weights and the inputs are bigger or equal to 0, so ReLU doesn't affect the products.

**Comment by alisidd** [comment]
@siryog90 My function is just equation (5) from the paper

**Comment by siryog90** [comment]
@alisidd are `a` and `b` assumed to be a single layer apart so then you can reduce this function over the whole graph?

**Comment by alisidd** [comment]
@siryog90 yep! Ideally the function would take in a list of tensors, which we could reduce ourselves.

**Comment by siryog90** [comment]
@alisidd makes sense. When I did the initial (incorrect) implementation, I used a list input as well, but then problem with a list is that defining the gradient of the regularizer is a bit weird. I think we should force the input to pack the weights into one higher rank tensor. We can unpack them within the `lp_path_regularizer`. This way, we can define the gradient with respect to a single, tensor input.

**Comment by AshishBora** [comment]
@alisidd  Can you confirm that your implementation is same as:

def lp_path_regularizer(a, b, p=2):
    a_p = pow(abs(a), p)
    b_p = pow(abs(b), p)
    return pow(reduce_sum(matmul(a_p, b_p)), 1.0/p)

**Comment by AshishBora** [comment]
Also, I think a more extensible approach can be to use a context manager.

All weight matrices defined within that context automatically have pseudo weight tensors created for them that are `pow(abs(weight), p)`. Every new bias will create pseudo bias of the same size but all zeros. Then every time a new tensor is created within that context, a pseudo tensor is created in parallel to mimic the operation but with pseudo weights and biases. (We need to be careful about keeping track of pseudo tensors and when to sue what). This essentially creates the copy of the tensorflow graph created within that context.

Now when you ask for a `Lp-PathNorm` between a and b: First check if `a` and `b` have pseudo counterparts. If not, then raise exception. If yes, then return

`sess.run(reduce_sum(b_pseudo, feed_dict={a_pseudo: ones_like(a)}))`

Going further, we would like this to be differentiable, so, not just returning a value but a tensor. So I guess there can be another function `make_parallel_graph(a, model_fn)`, where model_fn is a function that creates the model from, i.e. `b = model_fn(a)` without any path regularization.

This `make_parallel_graph` function should first create `a_pseudo = ones_like(a)`. Then call `model_fn` in the context defined above making parallel graph all the way. Finally it should return `(b, pow(reduce_sum(b_pseudo), 1.0/p))`,
where  the second term in the tuple is the basically the `Lp-PathNorm` we wanted. This is then a differentiable function of the weights and can added to the total loss or whatever.

I like this context manager based approach because it can work with convolutional networks or any other complex topology as well.

**Comment by siryog90** [comment]
@AshishBora On it. Will try and push a version for evaluation tomorrow evening.

**Comment by siryog90** [comment]
@AshishBora @alisidd What do you guys think about going in this direction? 
https://github.com/siryog90/tensorflow/commit/28ac4a10707542ff95380dc0eaa8933ff7f5a58f

The `Graph's` `create_op` function of a graph can then check if a `ContextGraph` is defined, and also call a `create_pseudo_op` of the `ContextGraph`.

**Comment by alisidd** [comment]
@AshishBora @siryog90 Wouldn't creating pseudo tensors for every tensor take up a lot of memory?

**Comment by siryog90** [comment]
@alisidd I am a bit hesitant to say how tf treats tensors in the memory. I will let @AshishBora respond in this. But one clear advantage of a pseudo graph is that the network no longer needs to be fully connected. Also, a pseudo network with the non linearities removed may be a  useful feature for other computations as well.

I have made some design updates, which I will push tonight.

**Comment by aselle** [comment]
@siryog90 ideally we'd have gradients in C++. We're still hoping to expose a version of tf.gradients in other languages than python.

**Comment by siryog90** [comment]
@alisidd @AshishBora Here is an update on the design. As @AshishBora suggested, to create what I call a "context graph", we use the `context_graph_scope` scope. The scope creates the pseudo graph and temporarily replaces the original graph's `create_op` function with one that in addition to everything `create_op` did, mutates the op and inserts it in the `context_graph`.

https://github.com/siryog90/tensorflow/commit/f8ee067129c9dcbadd629040114980ee28404e14

Thoughts?

**Comment by chivee** [comment]
Hi guys, does PathSGD have any following updates on the implementations?

**Comment by dynamicwebpaige** [comment]
PathNorm and PathSGD would be great adds as optimizers in [TensorFlow Addons](https://github.com/tensorflow/addons/tree/master/tensorflow_addons/optimizers), but are outside of the scope of the TensorFlow Core API.

Transferring this issue to `tensorflow/addons`; @seanpmorgan for visibility.
